### PR TITLE
DataBatch support gpu

### PR DESCRIPTION
**Pull request by cjfan**
To solve the issue #2770 
It pass my test both in cpu and gpu case, please check it.

### CODE REVIEW

File: python/mxnet/io.py [hunk 1]
```diff
@@ -411,10 +409,16 @@ def _getdata(self, data_source):
         """Load data from underlying arrays, internal use only"""
         assert(self.cursor < self.num_data), "DataIter needs reset."
         if self.cursor + self.batch_size <= self.num_data:
-            return [array(x[1][self.cursor:self.cursor+self.batch_size]) for x in data_source]
+            if isinstance(data_source[0][1], NDArray):
```
Comment (piiswrong) [inline python/mxnet/io.py]: This is not needed. array(mx.nd.ndarray) will still be ndarray.

File: python/mxnet/io.py [hunk 2]
```diff
@@ -411,10 +409,16 @@ def _getdata(self, data_source):
...
+                return [x[1][self.cursor:self.cursor+self.batch_size] for x in data_source]
+            else:
+                return [array(x[1][self.cursor:self.cursor+self.batch_size]) for x in data_source]
         else:
             pad = self.batch_size - self.num_data + self.cursor
-            return [array(np.concatenate((x[1][self.cursor:], x[1][:pad]),
+            if isinstance(data_source[0][1], NDArray):
+                return [concatenate([x[1][self.cursor:], x[1][:pad]]) for x in data_source]
+            else:
+                return [array(np.concatenate((x[1][self.cursor:], x[1][:pad]),
```
Comment (piiswrong) [inline python/mxnet/io.py]: same here. Change to
`return [concatenate([array(x[1][self.cursor:]), array(x[1][:pad])]) for x in data_source]`

**Comment by cjfan** [comment]
@piiswrong , thank you.   I thought array cound accept mx.nd.ndarray, but the current implementation does not support. 
```
File "D:\Work\Python27\lib\site-packages\mxnet-0.7.0-py2.7.egg\mxnet\io.py", line 403, in next
    return DataBatch(data=self.getdata(), label=self.getlabel(), \
  File "D:\Work\Python27\lib\site-packages\mxnet-0.7.0-py2.7.egg\mxnet\io.py", line 426, in getdata
    return self._getdata(self.data)
  File "D:\Work\Python27\lib\site-packages\mxnet-0.7.0-py2.7.egg\mxnet\io.py", line 415, in _getdata
    return [array(x[1][self.cursor:self.cursor+self.batch_size]) for x in data_source]
  File "D:\Work\Python27\lib\site-packages\mxnet-0.7.0-py2.7.egg\mxnet\ndarray.py", line 866, in array
    raise TypeError('source_array must be array like object')
TypeError: source_array must be array like object
```
the code in ndarray.py:
```
def array(source_array, ctx=None, dtype=mx_real_t):
    if not isinstance(source_array, np.ndarray):
        try:
            source_array = np.array(source_array, dtype=dtype)
        except:
            raise TypeError('source_array must be array like object')
    arr = empty(source_array.shape, ctx, dtype)
    arr[:] = source_array
    return arr
```
maybe can change to:
```
def array(source_array, ctx=None, dtype=mx_real_t):
    if isinstance(source_array, NDArray):  
        if ctx is None:
            ctx = source_array.context
    elif not isinstance(source_array, np.ndarray):  
##### following code no change
        try:
            source_array = np.array(source_array, dtype=dtype)
        except:
            raise TypeError('source_array must be array like object')
    arr = empty(source_array.shape, ctx, dtype)
    arr[:] = source_array
    return arr
```

but the concatenate in ndarray.py is same situation, can not satisfy both numpy and NDArray.
```
def concatenate(arrays, always_copy=True):
assert isinstance(arrays, list)
    assert len(arrays) > 0
    assert isinstance(arrays[0], NDArray)
```
I don't have a solution here.

on the other side, the pull request failed to pass some checks, 
such is the case, 
` if isinstance(data_source[0][1],NDArray):` 
will raise error when data_source is empty.
I have tested this, it can change to 
 ` if data_source and isinstance(data_source[0][1],NDArray):`

If there is any other suggestion, please tell me

**Comment by cjfan** [comment]
maybe, it can only store NDArray at _init_data() step.
because when _getdata(), it also will be change to NDArray.
if so, there will be no aboving trouble.

**Comment by cjfan** [comment]
fix the error of unittest to support shuffer.
now, DataBatch will storage all type of data in NDArray at the _init_data() step.
because the init is just one time, and the getdata() call if more frequently.
when use them , the data will aslo be given in NDArray type in the original implementation,
we can save the change step, as well NDArrayIter can support gpu now.
@piiswrong , please review it once more, thank you!

**Pull request merged**